go/test/endtoend/vtgate/vindex_bindvars/main_test.go

```diff
@@ -162,6 +187,28 @@ func TestMain(m *testing.M) {
 	os.Exit(exitCode)
 }
 
+func TestVindexHexTypes(t *testing.T) {
+	defer cluster.PanicHandler(t)
+	ctx := context.Background()
+	conn, err := mysql.Connect(ctx, &vtParams)
+	require.Nil(t, err)
+	defer conn.Close()
+
+	exec(t, conn, "INSERT INTO thex (id, field) VALUES "+
+		"(0x01,1), "+
+		"(x'a5',2), "+
+		"(0x48656c6c6f20476f7068657221,3), "+
+		"(x'c26caa1a5eb94096d29a1bec',4)")
+	result := exec(t, conn, "select id, field from thex order by id")
+
+	expected :=
+		"[[VARBINARY(\"\\x01\") INT64(1)] " +
+			"[VARBINARY(\"Hello Gopher!\") INT64(3)] " +
+			"[VARBINARY(\"\\xa5\") INT64(2)] " +
+			"[VARBINARY(\"\\xc2l\\xaa\\x1a^\\xb9@\\x96Қ\\x1b\\xec\") INT64(4)]]"
+	assert.Equal(t, expected, fmt.Sprintf("%v", result.Rows))
```

Comment:
@mattlord I've cherry picked your test from https://github.com/vitessio/vitess/pull/9616 here. I did have to change the expected output. Two notes:

1. Because of the context they're in, the hex values and numbers are always coerced to `VARBINARY` during evaluation, so their original types are lost. AFAICT this matches MySQL's behavior -- I hope it won't be a problem in practice.

2. Since the resulting rows are `order by id`, they should come as `1, 3, 2, 4`, since that's the proper ordering of the values.

Reply:
Internal transformations should be fine. I did a quick manual test with this branch and the results match up with [what I had seen with PlanValue before](https://github.com/vitessio/vitess/pull/9277):
```
$ mysql customer -e "select id, hex_keyspace_id, shard from customer.binary_vdx where id = 0x997901"
+------+-----------------+-------+
| id   | hex_keyspace_id | shard |
+------+-----------------+-------+
| �y  | 997901          | 80-   |
+------+-----------------+-------+
$ mysql customer -e "select id, hex_keyspace_id, shard from customer.binary_vdx where id = x'997901'"
+------+-----------------+-------+
| id   | hex_keyspace_id | shard |
+------+-----------------+-------+
| �y  | 997901          | 80-   |
+------+-----------------+-------+

$ mysql customer -e "select id, hex_keyspace_id, shard from customer.binary_md5_vdx where id = 0x997901"
+------+----------------------------------+-------+
| id   | hex_keyspace_id                  | shard |
+------+----------------------------------+-------+
| �y  | 2d2da332b59f56aeb235ecb36d639440 | -80   |
+------+----------------------------------+-------+
vitess@003f6cbe061b:/vt/local$ mysql customer -e "select id, hex_keyspace_id, shard from customer.binary_md5_vdx where id = x'997901'"
+------+----------------------------------+-------+
| id   | hex_keyspace_id                  | shard |
+------+----------------------------------+-------+
| �y  | 2d2da332b59f56aeb235ecb36d639440 | -80   |
+------+----------------------------------+-------+

$ mysql customer -e "select id, hex_keyspace_id, shard from customer.xxhash_vdx where id = 0x997901"
+------+------------------+-------+
| id   | hex_keyspace_id  | shard |
+------+------------------+-------+
| �y  | ac29dcaa4a0b4527 | 80-   |
+------+------------------+-------+
vitess@003f6cbe061b:/vt/local$ mysql customer -e "select id, hex_keyspace_id, shard from customer.xxhash_vdx where id = x'997901'"
+------+------------------+-------+
| id   | hex_keyspace_id  | shard |
+------+------------------+-------+
| �y  | ac29dcaa4a0b4527 | 80-   |
+------+------------------+-------+
```

THANK YOU! 😍